airflow/providers/amazon/aws/utils/eks_kube_config.py

```diff
@@ -0,0 +1,132 @@
+# software distributed under the License is distributed on an
+# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
+# KIND, either express or implied.  See the License for the
+# specific language governing permissions and limitations
+# under the License.
+import os
+from shutil import which
+from typing import Optional
+
+import boto3
+import yaml
+
+HOME = os.environ.get('HOME', '/tmp')
+DEFAULT_KUBE_CONFIG_FILENAME = 'config'
+DEFAULT_KUBE_CONFIG_PATH = str(os.path.join(HOME, '/.kube/', DEFAULT_KUBE_CONFIG_FILENAME))
+DEFAULT_CONTEXT_NAME = 'aws'
+DEFAULT_NAMESPACE_NAME = 'default'
+DEFAULT_POD_USERNAME = 'aws'
+
+
+def generate_config_file(
+    eks_cluster_name: str,
+    eks_namespace_name: str,
+    aws_profile: Optional[str],
+    kube_config_file_location: Optional[str] = DEFAULT_KUBE_CONFIG_PATH,
+    pod_username: Optional[str] = DEFAULT_POD_USERNAME,
+    pod_context: Optional[str] = DEFAULT_CONTEXT_NAME,
+    role_arn: Optional[str] = None,
+    aws_region: Optional[str] = None,
+) -> None:
+    """
+    Writes the kubeconfig file given an EKS Cluster name, AWS region, and file path.
+
+    :param eks_cluster_name: The name of the cluster to create the EKS Managed Nodegroup in.
+    :type eks_cluster_name: str
+    :param eks_namespace_name: The namespace to run within kubernetes.
+    :type eks_namespace_name: str
+    :param aws_profile: The named profile containing the credentials for the AWS CLI tool to use.
+    :type aws_profile: str
+    :param kube_config_file_location: Path to save the generated kube_config file to.
+    :type kube_config_file_location: str
+    :param pod_username: The username under which to execute the pod.
+    :type pod_username: str
+    :param pod_context: The name of the context access parameters to use.
+    :type pod_context: str
+    :param role_arn: The Amazon Resource Name (ARN) of the IAM role to associate with your nodegroup.
+    :type role_arn: str
+    :param aws_region: The name of the AWS Region the EKS Cluster resides in.
+    :type aws_region: str
+    """
+    installed = which("aws")
+    if installed is None:
+        message = (
+            "AWS CLI version 2 must be installed on the worker.  See: "
+            "https://docs.aws.amazon.com/cli/latest/userguide/install-cliv2.html"
+        )
+        print(message)
+        raise UnmetDependency(message)
+
+    # Set up the client
+    session = boto3.Session(region_name=aws_region, profile_name=aws_profile)
+    eks_client = session.client("eks")
+
+    # get cluster details
+    cluster = eks_client.describe_cluster(name=eks_cluster_name)
+    cluster_cert = cluster["cluster"]["certificateAuthority"]["data"]
+    cluster_ep = cluster["cluster"]["endpoint"]
+
+    # build the cluster config hash
+    cli_args = [
+        "--region",
+        aws_region,
+        "eks",
+        "get-token",
+        "--cluster-name",
+        eks_cluster_name,
+    ]
+    if role_arn:
+        cli_args.extend(["--role-arn", role_arn])
```

Comment:
Why we need to use it?

Reply:
Is it impossible to generate this token using STS?
```python
import base64
import boto3
import re
from botocore.signers import RequestSigner

STS_TOKEN_EXPIRES_IN = 60

def get_bearer_token(session, cluster_id, region):
    client = session.client('sts', region_name=region)
    service_id = client.meta.service_model.service_id

    signer = RequestSigner(
        service_id,
        region,
        'sts',
        'v4',
        session.get_credentials(),
        session.events
    )

    params = {
        'method': 'GET',
        'url': 'https://sts.{}.amazonaws.com/?Action=GetCallerIdentity&Version=2011-06-15'.format(region),
        'body': {},
        'headers': {
            'x-k8s-aws-id': cluster_id
        },
        'context': {}
    }

    signed_url = signer.generate_presigned_url(
        params,
        region_name=region,
        expires_in=STS_TOKEN_EXPIRES_IN,
        operation_name=''
    )

    base64_url = base64.urlsafe_b64encode(signed_url.encode('utf-8')).decode('utf-8')

    # remove any base64 encoding padding:
    return 'k8s-aws-v1.' + re.sub(r'=*', '', base64_url)

# If making a HTTP request you would create the authorization headers as follows:

session = boto3.session.Session()
headers = {'Authorization': 'Bearer ' + get_bearer_token(session, 'my_cluster', 'us-east-1')}
```

Reply:
In the case of Google, we should be using normal Google tokens sometime.

``hook.get_credentials().token`` should be enough to generate a valid access token, but it is not clearly documented in public docs.

Reply:
Here is a real-world example for usage gcloud access token to access kubernetes cluster:
https://registry.terraform.io/providers/hashicorp/google/latest/docs/guides/using_gke_with_terraform#using-the-kubernetes-and-helm-providers

Reply:
Using the AWS CLI was not our preference either, but had to settle for this due to other options not working correctly.  The EKS service does not support providing the token directly at all.  Searching online came up with a few workarounds, but the lesser evil was assuming that someone using an AWS service would be able to install an AWS tool.  There is also precedent for it in [Google's Kubernetes Operators](https://github.com/apache/airflow/blob/2625007c8aeca9ed98dea361ba13c2622482d71f/airflow/providers/google/cloud/operators/kubernetes_engine.py#L319) using the gcloud local script to accomplish the same thing.

I'll test using your provided example to see if that works and report back.  Dropping that dependency would be great.

Reply:
>  the lesser evil was assuming that someone using an AWS service would be able to install an AWS tool.

For me, this is very problematic for three reasons: 
- Passing credentials from Airflow to AWS CLI.
- AWS CLI is a native library because it cannot install the latest version via PIP, so in many cases it will be very difficult to use.
- Official Docker image doesn't have preinstalled AWS CLI.

Reply:
For more inspiration, you can also look at [aws-cli/awscli/customizations/eks/get_token.py](https://github.com/aws/aws-cli/blob/develop/awscli/customizations/eks/get_token.py), [kubergrunt eks token](https://github.com/gruntwork-io/kubergrunt), [kubernetes-sigs/aws-iam-authenticator](https://github.com/kubernetes-sigs/aws-iam-authenticator)

Reply:
Is the gcloud tool that the GKE operator uses in that link preinstalled in the official Docker image?  If so, could the AWS tool be as well?

Reply:
No. Gcloud is not preinstalled. We have a guide that explains how to add `gcloud` to a Docker image. See: http://airflow.apache.org/docs/docker-stack/recipes.html#google-cloud-sdk-installation

Reply:
Then I am missing what makes it ok for GKE users to have to install the gcloud cli tool but not for EKS users to have to install the aws cli tool.

Reply:
The KubernetesEngine operator requires an update. It was added when I didn't know a better way to authenticate. it should use google access-token for authentication: https://github.com/apache/airflow/pull/16571#discussion_r659255470

If you want, you can keep a dependency on AWS CLI (though I'd prefer not), but you definitely should pass credentials from Airflow to AWS CLI and that's the main issue. Now, this operator cannot be used on Cloud Composer, GKE, Astronomer, or many other environments that do not have set up the default AWS credentials in AWS CLI.

Reply:
The best way to integrate is when the token generation code would also use the Airflow hook because then we would have the greatest certainty that all credentials were passed from Airflow correctly. ;-)

Reply:
I still agree that it would be best to drop the dependency on the AWS CLI, I was just trying to understand the difference between the two cases... and that answers it. :)

My main concern with most of the alternatives/workarounds that I found online was that if something changes in the way STS vends the tokens then we're stuck with a broken product and trying to catch up.  We can (safely???) assume that the AWS CLI would always have the correct current means to get the token.  The AWS CLI code is open source, it's trivial to just copy/paste the relevant portions here, but that is asking for a problem, later IMHO.

I'm going to try to implement your suggested `get_bearer_token` code and see where that takes me.  I appreciate your patience explaining the issue.

Reply:
> that is asking for a problem, later IMHO.

I agree that this could bring unexpected problems in the future, but having to install AWS CLI by users and integrating Airflow with AWS CLI itself can be very difficult in edge cases too. We do not have a perfect solution here. The authorization method rarely changes compared to the product API, so I believe it will be enough stable approach.

Especially since it is described in the AWS documentation.
> Amazon EKS uses IAM to provide authentication to your Kubernetes cluster through [the AWS IAM authenticator for Kubernetes](https://github.com/kubernetes-sigs/aws-iam-authenticator). 


or

> This can be used as an alternative to the aws-iam-authenticator.

https://docs.aws.amazon.com/cli/latest/reference/eks/get-token.html

In the aws-iam-authenticator documentation, we have the code snippet I cited above.
https://github.com/kubernetes-sigs/aws-iam-authenticator#api-authorization-from-outside-a-cluster

Reply:
If you are concerned about maintaining this code, why not add it to AWS SDK for Python? It could have been helpful for other people as well. Have you thought to create an internal ticket on this topic?

Reply:
Is https://boto3.amazonaws.com/v1/documentation/api/latest/reference/services/eks.html#EKS.Client.generate_presigned_url of any help here?

Reply:
Early tests with the proposed get_bearer_token (https://github.com/apache/airflow/pull/16571#discussion_r659252727) are promising.  I'm going to play with some edge cases and see how it holds up, but I like the direction that is taking.

In general, are we happy with the security practices used there?  Any concerns we need to pay attention to or address with that approach?

Reply:
This should be addressed in https://github.com/apache/airflow/pull/16571/commits/a6ee2c369227212e12b8d5320c91fd63a6dc769d